src/java/org/apache/cassandra/index/sai/disk/v2/RowAwarePrimaryKeyFactory.java

```diff
@@ -0,0 +1,189 @@
+
+import org.apache.cassandra.db.Clustering;
+import org.apache.cassandra.db.ClusteringComparator;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+
+/**
+ * A row-aware {@link PrimaryKey.Factory}. This creates {@link PrimaryKey} instances that are
+ * sortable by {@link DecoratedKey} and {@link Clustering}.
+ */
+public class RowAwarePrimaryKeyFactory implements PrimaryKey.Factory
+{
+    private final ClusteringComparator clusteringComparator;
+
+    public RowAwarePrimaryKeyFactory(ClusteringComparator clusteringComparator)
+    {
+        this.clusteringComparator = clusteringComparator;
+    }
+
+    @Override
+    public PrimaryKey createTokenOnly(Token token)
+    {
+        return new RowAwarePrimaryKey(token, null, null, null);
+    }
+
+    @Override
+    public PrimaryKey createDeferred(Token token, Supplier<PrimaryKey> primaryKeySupplier)
+    {
+        return new RowAwarePrimaryKey(token, null, null, primaryKeySupplier);
+    }
+
+    @Override
+    public PrimaryKey create(DecoratedKey partitionKey, Clustering clustering)
+    {
+        return new RowAwarePrimaryKey(partitionKey.getToken(), partitionKey, clustering, null);
+    }
+
+    private class RowAwarePrimaryKey implements PrimaryKey
+    {
+        private Token token;
+        private DecoratedKey partitionKey;
+        private Clustering clustering;
+        private Supplier<PrimaryKey> primaryKeySupplier;
+
+        private RowAwarePrimaryKey(Token token, DecoratedKey partitionKey, Clustering clustering, Supplier<PrimaryKey> primaryKeySupplier)
+        {
+            this.token = token;
+            this.partitionKey = partitionKey;
+            this.clustering = clustering;
+            this.primaryKeySupplier = primaryKeySupplier;
+        }
+
+        @Override
+        public Token token()
+        {
+            return token;
+        }
+
+        @Override
+        public DecoratedKey partitionKey()
+        {
+            loadDeferred();
+            return partitionKey;
+        }
+
+        @Override
+        public Clustering clustering()
+        {
+            loadDeferred();
+            return clustering;
+        }
+
+        @Override
+        public PrimaryKey loadDeferred()
+        {
+            if (primaryKeySupplier != null && partitionKey == null)
+            {
+                PrimaryKey deferredPrimaryKey = primaryKeySupplier.get();
+                this.partitionKey = deferredPrimaryKey.partitionKey();
+                this.clustering = deferredPrimaryKey.clustering();
+                primaryKeySupplier = null;
+            }
+            return this;
+        }
+
+        @Override
+        public ByteSource asComparableBytes(ByteComparable.Version version)
```

Comment:
This loads all the components of the key. I wonder, in what contexts we do not compare keys by this method, and use `compareTo` instead.

Looks like `compareTo` is potentially faster because it might not need to load the primary key at all, if the tokens are different. Is it correct? 

 I mean - it is a pitty the lazy key loading optimization does not work with this method.

Nothing to do here, just checking if I understand it correctly.

Reply:
So this method is used in 2 situations. First during writing when it is always fully loaded, the second is when the key is being used to look up a rowId from the map. In the second case the key has to be fully loaded otherwise it may not directly match the key in the map. This showed up in testing where tests were failing because the key was not fully loaded.